### ISSUE TITLE
Perf Remote Target Environment

### ISSUE THREAD

**Issue by tustvold** [influxdata/influxdb_iox#3143]
**Problem**

As part of developing IOx we wish to be able to reproduce or simulate specific workloads in a controlled environment, and potentially incorporate this into a performance regression suite run by CI.

We have the beginnings of this in the form of the perf scripts written by @jacobmarble. However, these are currently limited to running tests on a local machine, which has a couple of implications:

* Performance characteristics may vary across different hardware / OSes
* Cannot easily test the impact of network/disk IO
* Limited to the resources of a single machine
* May be impractical to leave "soaking" whilst getting on with other work
* Cannot easily test different deployment topologies

**Proposal**

Extend `perf` with the concept of a target environment. This would replace the current command line flags that configure aspects such as what object store implementation to use. Instead `perf` would be provided a `toml` file that identifies a deployment target, and provides configuration options specific to that target. 

One deployment target would be `docker-local` which would use the same combination of docker and `./target` currently and with the same options that are currently exposed via CLI.

Another deployment target would be `kustomize` which would take a set of k8s manifests, and use [kustomize](https://kustomize.io/) to apply them to a target cluster. _I don't have a particular affinity for kustomize, I chose it simply because it is built-in to kubectl, it's the flexibility of using a templating tool that matters to me, if people would prefer using jsonnet or helm or whatever_ :shrug:

The expectation would be to end up with a collection of target environment configs checked into git, allowing test batteries to be run against a number of different toplogies and environments, depending on what the developer is trying to simulate/reproduce.

This would also provide an obvious extension point for additional target environments, e.g. `local`, or `podman` or `k8s-jsonnet` or `helm` or `conductor` if we so wished.

**Comment by e-dard** [comment]
I think that we should probably try and align effort around Remocal. Jumping into those details might be best done out of band of the public tracker.

I certainly like the idea of having multiple targets 👍

**Comment by alamb** [comment]
I think it may also be worth considering splitting up the two concerns into two sets of tools:
1. Setting up a IOx cluster/clusters
2. Running a battery of tests (perhaps driving the `data_generator` that @pauldix  has been working on)

Today `perf.py` does both of these pieces -- splitting it up into two (or more) tools might make it easier to customize rather than have one tool that does it all

**Comment by tustvold** [comment]
> I think that we should probably try and align effort around Remocal.

I was hoping to keep the scope down and avoid that particular can of worms by just focusing on orchestrating IOx and not the whole shebang, but happy to discuss.

> I think it may also be worth considering splitting up the two concerns into two sets of tools:

Perhaps, although said data generator(s) need to run somewhere and so need to be orchestrated by something... Along with any supporting services to collect metrics, traces, etc... That something could be a different tool though for sure, but I kind of like the single entrypoint...

**Comment by crepererum** [comment]
@mkmik and I had a chat and came to a similar conclusion:

1. we need small tools/puzzle pieces (e.g. "deploy a simple IOx" cluster and "run a whole test battery" are two tools, where the 2nd one probably uses the first)
2. since tools are simple and configurable, they can be modified / configured (e.g. do the whole experiment on a remocal K8s cluster or locally or even in some completely different cluster that you already have)
3. we start small and add features over time (e.g. the data generator)
4. the whole thing is available for OSS contributors, but we might have some secret influxdata sauce somewhere to simplify the interaction w/ our own test clusters

**Comment by pauldix** [comment]
I think Perf was designed specifically to just run on laptop as a lightweight thing. For a real environment (i.e. k8s), we have a tool for "deploy an IOx cluster", which is Conductor. Do we want some other script that manually creates different topologies?

The Data Generator doesn't run a whole test battery, it just generates data however you specify, which can run to completion or continue running forever. So it wouldn't qualify as the tool that can run a test battery.

I'd like to be able to run a command to get a k8s environment set up (I believe this is remocal). I believe this also includes the entire Cloud environment + Conductor + Kafka, etc. Ideally this will also include everything for gathering telemetry.

Command to run a single Data Generator spec in the target environment. So this would spin up a pod and start the DG, which will call to Conductor to create the database (or many databases depending on the spec) and then start generating load.

Command to run a battery of tests (this would be some dir with a Data Generator configs and queries, etc). We probably need to write a separate tool like a Query Generator that could create query load, verify correctness(?).

**Comment by tustvold** [comment]
> Perf was designed specifically to just run on laptop as a lightweight thing

I specified the ticket in terms of perf purely because it exists, but there seems to be consensus around splitting the concerns, which is fine with me

> Do we want some other script that manually creates different topologies?

Yes. Different topologies, environments, load generators, write buffer implementations (Kafka vs Redpanda), object store implementations (minio vs S3 vs ...), networking (Istio vs no-Istio), etc... The selling point of k8s is being able to describe all manner of random things in a set of manifests and anyone be able to deploy it. 

I was envisaging this "script" would pretty much just be `kubectl apply -k`, with a little bit of setup and teardown logic. It needn't be anything particularly sophisticated :grin: 

>  run a command to get a k8s environment set up (I believe this is remocal)

TBC I was taking having a k8s namespace as a given here, I wasn't proposing scripting that :laughing:

**Comment by alamb** [comment]
> I was envisaging this "script" would pretty much just be kubectl apply -k, with a little bit of setup and teardown logic. It needn't be anything particularly sophisticated 😁

This is basically what `make setup-remocal` does -- of couse what is being `apply -f` I would say is "non trivial" 😆 

I would be happy to do a 🧠 dump on anyone who wants to see what remocal does today (and what I hope it will do 🔜 )

**Comment by tustvold** [comment]
> of couse what is being apply -f I would say is "non trivial"

Indeed, it is highly non-trivial and highly coupled into how the cloud platform works and is deployed. This ticket proposes something far simpler that just spins up an IOx deployment in any old k8s cluster and applies some load to it. This may or may not match the current cloud configuration, because the whole purpose is to be able to experiment with configurations that are different from the current blessed cloud topology :laughing:. One could even argue that the presence of load generators is a divergence from the cloud topology...

To phrase it differently, having to spin up a complete replica of our cloud platform in order to run some distributed benchmarks against IOx seems not only a tad excessive, but unnecessarily limiting to what the cloud platform does/requires? By contrast what this proposes is just a couple of yaml files containing k8s manifests, that can be deployed by anyone to any k8s cluster - be that remocal, minikube, a k3s cluster running on a collection of raspberry pis, a random GKE cluster, it doesn't matter :grin:  

> I would be happy to do a brain dump on anyone

Let's discuss in Monday morning standup?
